api/v4/common_types.go

```diff
@@ -348,9 +361,39 @@ type AppSourceDefaultSpec struct {
 	// +optional
 	VolName string `json:"volumeName,omitempty"`
 
-	// Scope of the App deployment: cluster, clusterWithPreConfig, local. Scope determines whether the App(s) is/are installed locally or cluster-wide
+	// Scope of the App deployment: cluster, clusterWithPreConfig, local, premiumApps. Scope determines whether the App(s) is/are installed locally, cluster-wide or its a premium app
 	// +optional
 	Scope string `json:"scope,omitempty"`
+
+	// Properties for premium apps, fill in when scope premiumApps is chosen
+	// +optional
+	PremiumAppsProps PremiumAppsProps `json:"premiumAppsProps,omitempty"`
+}
+
+// PremiumAppsProps represents properties for premium apps such as ES
+type PremiumAppsProps struct {
+	// Type: enterpriseSecurity for now, can accomodate itsi etc.. later
+	// +optional
+	Type string `json:"type,omitempty"`
+
+	// Enterpreise Security App defaults
+	// +optional
+	EsDefaults EsDefaults `json:"esDefaults,omitempty"`
+}
+
+// EsDefaults captures defaults for the Enterprise Security App
+type EsDefaults struct {
+	// Sets the sslEnablement value for ES app installation
+	//     strict: Ensure that SSL is enabled
+	//             in the web.conf configuration file to use
+	//             this mode. Otherwise, the installer exists
+	//	     	   with an error.
+	//     auto: Enables SSL in the etc/system/local/web.conf
+	//           configuration file. This is the DEFAULT mode
+	//           used by the operator if left empty.
+	//     ignore: Ignores whether SSL is enabled or disabled.
+	// +optional
+	SslEnablement string `json:"sslEnablement,omitempty"`
```

Comment:
How about something like this?? - 
```
type GenericDefaults interface{}

type PremiumAppsDefaults interface {
	GetDefaults() GenericDefaults
}

// PremiumAppsProps represents properties for premium apps such as ES
type PremiumAppsProps struct {
	// Type: enterpriseSecurity for now, can accomodate itsi etc.. later
	// +optional
	Type string `json:"type,omitempty"`

	PremiumAppsDefaults PremiumAppsDefaults `json:"premiumAppsDefaults,omitempty"`
}

type EsDefaultsHandler struct {
	// Enterpreise Security App defaults
	// +optional
	EsDefaults EsDefaults `json:"esDefaults,omitempty"`
}

var _ GenericDefaults = &EsDefaults{}

// EsDefaults captures defaults for the Enterprise Security App
type EsDefaults struct {
	// Sets the sslEnablement value for ES app installation
	//     strict: Ensure that SSL is enabled
	//             in the web.conf configuration file to use
	//             this mode. Otherwise, the installer exists
	//	     	   with an error.
	//     auto: Enables SSL in the etc/system/local/web.conf
	//           configuration file. This is the DEFAULT mode
	//           used by the operator if left empty.
	//     ignore: Ignores whether SSL is enabled or disabled.
	// +optional
	SslEnablement string `json:"sslEnablement,omitempty"`
}

func (EsDefaultsHandler *EsDefaultsHandler) GetDefaults() GenericDefaults {

	return EsDefaultsHandler.EsDefaults
}
```
This would make it more generic and in the future, implementation of other future premium apps such as ITSI can be easily integrated.

Reply:
I guess the use of interfaces is not supported while generating CRD spec, hence parking this approach for now.

Reply:
Per our discussion, controller-gen throws the following error when using interface as a type:

```
/Users/akondur/Desktop/operator/test/splunk-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
github.com/splunk/splunk-operator/api/v4:-: invalid type: interface{}
github.com/splunk/splunk-operator/api/v4:-: invalid type: interface{GetDefaults() github.com/splunk/splunk-operator/api/v4.GenericDefaults}
Error: not all generators ran successfully
run `controller-gen object:headerFile=hack/boilerplate.go.txt paths=./... -w` to see all available markers, or `controller-gen object:headerFile=hack/boilerplate.go.txt paths=./... -h` for usage
make: *** [generate] Error 1
I 
```